File: solutions/csharp/allergies/1/Allergies.cs
using System;
using System.Collections.Generic;

public enum Allergen : int
{
    Eggs,
    Peanuts,
    Shellfish,
    Strawberries,
    Tomatoes,
    Chocolate,
    Pollen,
    Cats
}

public class Allergies
{
    private HashSet<Allergen> _allergies = new();

    public Allergies(int mask)
    {
        foreach (Allergen a in Enum.GetValues(typeof(Allergen)))
        {
            int bitResult = (1 << (int)a) & mask;
            if (bitResult > 0)
                _allergies.Add(a);
        }
    }

    public bool IsAllergicTo(Allergen allergen) => _allergies.Contains(allergen);

    public Allergen[] List()
    {
        Allergen[] allergenList = new Allergen[_allergies.Count];
        _allergies.CopyTo(allergenList);

        return allergenList;
    }
}

File: solutions/cpp/allergies/1/allergies.cpp
#include "allergies.h"
#include <unordered_set>

namespace allergies {
allergy_test::allergy_test(unsigned int allergy_score) : allergies() {
  set_allergies(allergy_score);
};

bool allergy_test::is_allergic_to(std::string item) {
  return allergies.find(item) != allergies.end();
}

std::unordered_set<std::string> allergy_test::get_allergies() {
  return allergies;
}

void allergy_test::set_allergies(int allergy_score) {
  if (allergy_score & static_cast<int>(Items::Eggs)) {
    allergies.emplace("eggs");
  }
  if (allergy_score & static_cast<int>(Items::Peanuts)) {
    allergies.emplace("peanuts");
  }
  if (allergy_score & static_cast<int>(Items::Shellfish)) {
    allergies.emplace("shellfish");
  }
  if (allergy_score & static_cast<int>(Items::Shellfish)) {
    allergies.emplace("shellfish");
  }
  if (allergy_score & static_cast<int>(Items::Strawberries)) {
    allergies.emplace("strawberries");
  }
  if (allergy_score & static_cast<int>(Items::Tomatoes)) {
    allergies.emplace("tomatoes");
  }
  if (allergy_score & static_cast<int>(Items::Chocolate)) {
    allergies.emplace("chocolate");
  }
  if (allergy_score & static_cast<int>(Items::Pollen)) {
    allergies.emplace("pollen");
  }
  if (allergy_score & static_cast<int>(Items::Cats)) {
    allergies.emplace("cats");
  }
}

} // namespace allergies
